semantics-3.0/Semantics.hs

```diff
@@ -236,206 +227,219 @@ data ReduceWarning = ReduceNoWarning
                                     -- ^ src    ^ dest ^ paid ^ expected
                    | ReduceShadowing ValueId Integer Integer
                                      -- oldVal ^  newVal ^
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
 
-data ReduceEffect = ReduceNoEffect
-                  | ReduceNormalPay Party Money
-  deriving (Eq,Ord,Show,Read)
+data Payment = Payment Party Money
+  deriving (Eq,Ord,Show)
 
-data ReduceError = ReduceAmbiguousSlotInterval
-  deriving (Eq,Ord,Show,Read)
+type ReduceEffect = Maybe Payment
 
 data ReduceResult = Reduced ReduceWarning ReduceEffect State Contract
                   | NotReduced
-                  | ReduceError ReduceError
-  deriving (Eq,Ord,Show,Read)
+                  | AmbiguousSlotIntervalError
+  deriving (Eq,Ord,Show)
 
 
 -- | Carry a step of the contract with no inputs
-reduce :: Environment -> State -> Contract -> ReduceResult
-reduce env state contract = case contract of
-    Refund -> case refundOne $ accounts state of
-        Just ((party, money), newAccounts) ->
-            let newState = state { accounts = newAccounts }
-            in  Reduced ReduceNoWarning (ReduceNormalPay party money) newState Refund
+reduceContractStep :: Environment -> State -> Contract -> ReduceResult
+reduceContractStep env state contract = case contract of
+
+    Refund -> case refundOne (accounts state) of
+        Just ((party, money), newAccounts) -> let
+            newState = state { accounts = newAccounts }
+            in Reduced ReduceNoWarning (Just (Payment party money)) newState Refund
         Nothing -> NotReduced
-    Pay accId payee val cont -> if amountToPay <= 0
-        then Reduced (ReduceNonPositivePay accId payee amountToPay) ReduceNoEffect state cont
-        else let
-                (paidMoney, newAccs) = withdrawMoneyFromAccount accId amountToPay $ accounts state
-                paidAmount = paidMoney
-                warning = if paidAmount < amountToPay
-                          then ReducePartialPay accId payee paidAmount amountToPay
+
+    Pay accId payee val cont -> let
+        amountToPay = evalValue env state val
```

Comment:
does it make sense for `evalValue` to return `Ada` instead of `Integer`?

Reply:
+1

I doubt, because it may be a block number. Also it may be a different currency eventually.

This could possibly be a GADT like Simon suggested, if we want to give it a specific type. We can think about that in the future.